server/linux_x11/server_x11.py

```diff
@@ -211,131 +206,166 @@ def write_command(message, arguments='type --file -', executable='xdotool'):
-       window. it is fine to include platform specific information, but
-       at least include title and executable.'''
 
-    flush_xdotool(_xdotool)
-    window_id, window_title = get_active_window()
-    if window_id is None:
+def get_context():
+    """
+    Query the system for context information.  This data is typically passed
+    back to the aenea client so that it may use it in Dragonfly grammars.
+    Specifically, this data will be used when Dragonfly's grammars perform
+    context matching to decide which grammars should be activated.
+    :return: various properties related to the current active window
+      All possible return keys are demonstrated below:
+
+      {
+          'id': 1234,             # window id
+          'cls_name': 'foo',      # WM_CLASS name
+          'cls': 'bar',           # WM_CLASS
+          'title': 'baz',         # WM_NAME
+          'client_machine': 'me', # WM_CLIENT_MACHINE
+          'name': 'baz',          # WM_NAME,
+          'locale': 'x',          # WM_LOCALE_NAME
+          'pid': 123,             # _NET_WM_PID
+          'type': '543',          # x11 atom for _NET_WM_WINDOW_TYPE
+          'role': '...',          # WM_WINDOW_ROLE
+          'desktop': '0',         # _NET_WM_DESKTOP
+      }
+    :rtype dict
+    """
+    try:
+        window_id = libxdo.get_focused_window_sane()
+        window = display.create_resource_object('window', window_id)
+    except Exception as error:
+        logger.error('failed to get active window error=%s', error)
         return {}
 
-    properties = {
-        'id': window_id,
-        'title': window_title,
-        }
-    for line in read_command('-id %s' % window_id, 'xprop').split('\n'):
-        split = line.split(' = ', 1)
-        if len(split) == 2:
-            rawkey, value = split
-            if split[0] in _XPROP_PROPERTIES:
-                property_value = value[1:-1] if '(STRING)' in rawkey else value
-                properties[_XPROP_PROPERTIES[rawkey]] = property_value
-            elif rawkey == 'WM_CLASS(STRING)':
-                window_class_name, window_class = value.split('", "')
-                properties['cls_name'] = window_class_name[1:]
-                properties['cls'] = window_class[:-1]
-
-    # Sigh...
-    properties['executable'] = None
-    properties['executable'] = None
-    if 'pid' in properties:
-        try:
-            proc_command = '/proc/%s/exe' % properties['pid']
-            properties['executable'] = os.readlink(proc_command)
-        except OSError:
-            ps = read_command('%s' % properties['pid'], executable='ps')
-            ps = ps.split('\n')[1:]
-            if ps:
-                try:
-                    properties['executable'] = ps[0].split()[4]
-                except Exception:
-                    pass
+    properties = {'id': window_id}
+
+    window_class = window.get_wm_class()
+    if window_class is not None:
+        properties['cls_name'] = window_class[1]
```

Comment:
this flips cls_name and cls from the old server. I initially misunderstood this thinking that it didn't work.

I was looking around to see if it's this version that's wrong or the old version was wrong; I'm not convinced the old way is right. Did you mean to flip these? Which way is correct?